File: fs-core/src/main/java/xyz/srclab/common/io/RandomOutputStream.java
package xyz.srclab.common.io;

import xyz.srclab.common.base.FsCheck;

import java.io.IOException;
import java.io.OutputStream;
import java.io.RandomAccessFile;

final class RandomOutputStream extends OutputStream {

    private final RandomAccessFile random;
    private final long limit;
    private long pos;

    RandomOutputStream(RandomAccessFile random, long offset, long length) {
        try {
            if (length != -1) {
                FsCheck.checkArgument(offset >= 0 && length >= 0, "offset and length must >= 0.");
                this.limit = offset + length;
            } else {
                this.limit = length;
            }
            this.random = random;
            this.pos = offset;
            this.random.seek(pos);
        } catch (IOException e) {
            throw new FsIOException(e);
        }
    }

    @Override
    public synchronized void write(byte[] b, int off, int len) throws IOException {
        try {
            FsCheck.checkRangeInBounds(off, off + len, 0, b.length);
            if (limit != -1) {
                FsCheck.checkInBounds(pos + len - 1, pos, limit);
            }
            random.write(b, off, len);
            pos += len;
        } catch (IOException e) {
            throw e;
        } catch (Exception e) {
            throw new IOException(e);
        }
    }

    @Override
    public synchronized void write(byte[] b) throws IOException {
        write(b, 0, b.length);
    }

    @Override
    public synchronized void write(int b) throws IOException {
        try {
            if (limit != -1) {
                FsCheck.checkInBounds(pos, pos, limit);
            }
            random.write(b);
            pos += 1;
        } catch (IOException e) {
            throw e;
        } catch (Exception e) {
            throw new IOException(e);
        }
    }

    @Override
    public void flush() throws IOException {
        random.getChannel().force(true);
    }

    @Override
    public void close() throws IOException {
        random.close();
    }
}

File: boat-core/src/main/kotlin/xyz/srclab/common/io/RandomOutputStream.kt
package xyz.srclab.common.io

import xyz.srclab.common.base.checkInBounds
import xyz.srclab.common.base.checkRangeInBounds
import java.io.OutputStream
import java.io.RandomAccessFile

/**
 * Makes [RandomAccessFile] as destination of [OutputStream].
 */
open class RandomOutputStream(
    private val destination: RandomAccessFile,
    private val offset: Long,
    private val length: Long
) : OutputStream() {

    init {
        checkRangeInBounds(offset, offset + length, 0, destination.length())
        destination.seek(offset)
    }

    private var pos = offset

    override fun write(b: Int) {
        pos.checkInBounds(offset, offset + length)
        destination.write(b)
        pos++
    }

    override fun write(b: ByteArray, off: Int, len: Int) {
        checkRangeInBounds(off, off + len, 0, b.size)
        val nextPos = pos + len
        (nextPos - 1).checkInBounds(offset, offset + length)
        destination.write(b, off, len)
        pos = nextPos
    }
}